SpatialDeconvolution.cu

```diff
@@ -0,0 +1,369 @@
+  // and always contains ones.
+  if (ones->nDimension != 2 || ones->size[0]*ones->size[1] < outputHeight*outputWidth) {
+    // Resize plane and fill with ones...
+    THCudaTensor_resize2d(state, ones, outputHeight, outputWidth);
+    THCudaTensor_fill(state, ones, 1);
+  }
+
+  // Helpers
+  THCudaTensor *input_n = THCudaTensor_new(state);
+  THCudaTensor *output_n = THCudaTensor_new(state);
+
+  // For each elt in batch, do:
+  for (int elt = 0; elt < batchSize; elt ++) {
+    // Matrix mulitply per output:
+    THCudaTensor_select(state, input_n, input, 0, elt);
+    THCudaTensor_select(state, output_n, output, 0, elt);
+
+    // M,N,K are dims of matrix A and B
+    // (see http://docs.nvidia.com/cuda/cublas/#cublas-lt-t-gt-gemm)
+    long m = weight->size[1];
+    long n = columns->size[1];
+    long k = weight->size[0];
+
+    // Do GEMM (note: this is a bit confusing because gemm assumes column-major matrices)
+    THCudaBlas_gemm(
+        state,
+        'n', 't',
+        n, m, k,
+        1,
+        THCudaTensor_data(state, input_n), n,
+        THCudaTensor_data(state, weight), m,
+        0,
+        THCudaTensor_data(state, columns), n
+    );
+
+    // Unpack columns back into input:
+    col2im(
+      THCState_getCurrentStream(state),
+      THCudaTensor_data(state, columns),
+      nOutputPlane, outputHeight, outputWidth, kH, kW, padH, padW, dH, dW,
+      THCudaTensor_data(state, output_n)
+    );
+
+    // Do Bias after:
+    // M,N,K are dims of matrix A and B
+    // (see http://docs.nvidia.com/cuda/cublas/#cublas-lt-t-gt-gemm)
+    long m_ = nOutputPlane;
+    long n_ = outputHeight * outputWidth;
+    long k_ = 1;
+
+    // Do GEMM (note: this is a bit confusing because gemm assumes column-major matrices)
+    THCudaBlas_gemm(
+        state,
+        't', 'n',
+        n_, m_, k_,
+        1,
+        THCudaTensor_data(state, ones), k_,
+        THCudaTensor_data(state, bias), k_,
+        1,
+        THCudaTensor_data(state, output_n), n_
+    );
+
+  }
+
+  // Free
+  THCudaTensor_free(state, input_n);
+  THCudaTensor_free(state, output_n);
+
+  // Resize output
+  if (batch == 0) {
+    THCudaTensor_resize3d(state, output, nOutputPlane, outputHeight, outputWidth);
+    THCudaTensor_resize3d(state, input, nInputPlane, inputHeight, inputWidth);
+  }
+
+  // return output
+  return 1;
+}
+
+static int cunn_SpatialDeconvolution_updateGradInput(lua_State *L) {
+  THCState *state = getCutorchState(L);
+  // Inputs
+  THCudaTensor *input = (THCudaTensor *)luaT_checkudata(L, 2, "torch.CudaTensor");
+  THCudaTensor *gradOutput = (THCudaTensor *)luaT_checkudata(L, 3, "torch.CudaTensor");
+
+  // Params
+  int dW = luaT_getfieldcheckint(L, 1, "dW");
+  int dH = luaT_getfieldcheckint(L, 1, "dH");
+  int kW = luaT_getfieldcheckint(L, 1, "kW");
+  int kH = luaT_getfieldcheckint(L, 1, "kH");
+  int nOutputPlane = luaT_getfieldcheckint(L, 1, "nInputPlane");
```

Comment:
What is the idea behind reversing the meaning of the nInputPlane and nOutputPlane? This is not 'compatible' with @fmassa 's tests in https://github.com/fmassa/nn/blob/deconv/test.lua#L1658 (e.g. in 1670, 1671)